src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/ApplyVocabularyMappings.java

```diff
@@ -67,14 +47,44 @@ public static LogicalPlan rewriteToUseLocalVocabulary( final LogicalPlan inputPl
 	public LogicalPlan apply( final LogicalPlan inputPlan ) {
 		if (inputPlan.getRootOperator() instanceof LogicalOpRequest) {
 			final LogicalOpRequest<?,?> requestOp = (LogicalOpRequest<?,?>) inputPlan.getRootOperator();
-			if(requestOp.getFederationMember().getVocabularyMapping() != null) { // If fm has a vocabulary mapping vm
-				final LogicalOpLocalToGlobal l2g = new LogicalOpLocalToGlobal(requestOp.getFederationMember().getVocabularyMapping());
-				final LogicalPlan rw = rewriteToUseLocalVocabulary(inputPlan);
-				return new LogicalPlanWithUnaryRootImpl(l2g,rw);
-			} else {
+
+			// For a compressed version of query plan, check if it is necessary to add the l2g operator over a request based on the form of the graph pattern.
+			// The current implementation assumes that only concepts and roles are being considered in vocabulary mapping.
+			boolean rewriteNeeded = false;
+			if ( compressVocabularyRewriting ) {
+				final Set<TriplePattern> tps = LogicalOpUtils.getTriplePatternsOfReq(requestOp);
+				if ( !tps.isEmpty() ) {
+					for ( final TriplePattern tp : tps ) {
+						// If any triple pattern is in the form of (-, ?p, -) or (-, rdfs:type, ?o),
+						// the intermediate results might need to be rewritten. This requires adding a L2G operator over the request.
+						if ( tp.asJenaTriple().getPredicate().isVariable()
+								|| (tp.asJenaTriple().getPredicate().hasURI("<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>")

```

Comment:
Instead of hard-coding the URI of rdf:type as a string, please use the class `org.apache.jena.vocabulary.RDF`.